tests/firecrackerdrivertest.c

```diff
@@ -0,0 +1,302 @@
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_RUNNING);
+
+    return 0;
+}
+
+static int
+testShutdownDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 500;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainShutdown(dom) < 0) {
+        return -1;
+    }
+
+    do {
+        g_usleep(50 * 1000);
+
+        if (virDomainGetState(dom, &state, NULL, 0) < 0) {
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_SHUTOFF);
+
+    return 0;
+}
+
+static int
+testUndefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainUndefine(dom) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCloseConnection(const void *opaque G_GNUC_UNUSED)
+{
+    if (virConnectClose(test_conn) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+# define MY_SUBTEST(TESTFUNC, arg) \
+    do { \
+        fprintf(stderr, #TESTFUNC "   ... "); \
+        if (TESTFUNC(opaque) < 0) { \
+            fprintf(stderr, "FAIL"); \
+            return -1; \
+        } else { \
+            fprintf(stderr, "OK"); \
+        } \
+        fprintf(stderr, "\n"); \
+    } while (0)
+
+static int
+testLifecycle(const void *opaque G_GNUC_UNUSED)
+{
+    fprintf(stderr, "\n");
+
+    MY_SUBTEST(testInitiateConnection, opaque);
+    MY_SUBTEST(testDefineDomain, opaque);
+    MY_SUBTEST(testCreateDomain, opaque);
+    MY_SUBTEST(testShutdownDomain, opaque);
+    MY_SUBTEST(testUndefineDomain, opaque);
+    MY_SUBTEST(testCloseConnection, opaque);
+
+    return 0;
+}
+
+static int
+testDefineShutdown(const void *opaque G_GNUC_UNUSED)
+{
+    fprintf(stderr, "\n");
+
+    MY_SUBTEST(testInitiateConnection, opaque);
+    MY_SUBTEST(testDefineDomain, opaque);
+
+    if (testShutdownDomain(opaque) == 0) {
+        fprintf(stderr, "Should not be able to shutdown only defined domain \n");
+        return -1;
+    }
+
+    MY_SUBTEST(testUndefineDomain, opaque);
+    MY_SUBTEST(testCloseConnection, opaque);
+
+    return 0;
+}
+
+# define DO_TEST(TESTNAME, NAME) \
+    if (virTestRun(TESTNAME, \
+                   NAME, NULL) < 0) \
+        ret = -1
+
+static int
+mymain(void)
+{
+    int ret = 0;
+
+    DO_TEST("Open and close driver connection ", testDriverConnection);
+
+    downloadKernelAndRootFS();
+    DO_TEST("Lifecycle", testLifecycle);
+
+    DO_TEST("Shutdown after define -> SHUT_OFF state", testDefineShutdown);
```

Comment:
say that this is a negative test

Reply:
I've modified the comment, did you have something else in mind?

Reply:
yep, that's what I meant